README.md
# kixi-metrics

test/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-02-23
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-02-23
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/kixi.metrics/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/kixi.metrics/compare/0.1.0...0.1.1


test/project.clj
(def metrics-version "2.9.0")
(defproject kixi.metrics "0.1.0-SNAPSHOT"
  :description "Provides mappers for jvm metrics to json and a Reporter for standard out"
  :url "http://github.com/MastodonC/kixi.metrics"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [cheshire "5.7.0"]
                 [metrics-clojure ~metrics-version]
                 [metrics-clojure-jvm ~metrics-version]])


test/kixi/metrics_test.clj
(ns kixi.metrics-test
  (:require [clojure.test :refer :all]
            [kixi.metrics :refer :all]
            [metrics
             [core :as m :refer [default-registry]]
             [counters :as c]
             [gauges :as g]
             [timers :as t]
             [histograms :as h]
             [meters :as meter]]
            [metrics.jvm.core :as jvm])
  (:import [com.codahale.metrics Counter]))

(deftest instrument-jvm-check-metrics
  (let [reg (m/new-registry)
        _ (jvm/instrument-jvm reg)
        metric-maps (registry->maps reg)]
    (is (= (count (filter :log-type metric-maps))
           (count metric-maps)))))

(def zero-counter
  {:type :counter
   :name "default.default.a-counter"
   :value 0})

(deftest map-counter->map
  (let [reg (m/new-registry)
        counter (c/counter reg "a-counter")
        cs (map counter->map (m/counters reg))]
    (is (= zero-counter
           (first cs)))))

(def zero-gauge
  {:type :gauge
   :name "default.default.a-gauge"
   :value 10})

(deftest map-gauge->map
  (let [reg (m/new-registry)
        gauge (g/gauge-fn reg "a-gauge" (constantly 10))
        gs (map gauge->map (m/gauges reg))]
    (is (= zero-gauge
           (first gs)))))

(def zero-timer
  [{:type :timer, :name "default.default.a-timer.min", :value 0}
   {:type :timer, :name "default.default.a-timer.p50", :value 0.0}
   {:type :timer, :name "default.default.a-timer.p95", :value 0.0}
   {:type :timer, :name "default.default.a-timer.mean", :value 0.0}
   {:type :timer, :name "default.default.a-timer.p98", :value 0.0}
   {:type :timer, :name "default.default.a-timer.count", :value 0}
   {:type :timer, :name "default.default.a-timer.p99", :value 0.0}
   {:type :timer, :name "default.default.a-timer.std-dev", :value 0.0}
   {:type :timer, :name "default.default.a-timer.max", :value 0}
   {:type :timer, :name "default.default.a-timer.p75", :value 0.0}
   {:type :timer, :name "default.default.a-timer.p999", :value 0.0}
   {:type :timer, :name "default.default.a-timer.m1_rate", :value 0.0}
   {:type :timer, :name "default.default.a-timer.m5_rate", :value 0.0}
   {:type :timer, :name "default.default.a-timer.m15_rate", :value 0.0}
   {:type :timer, :name "default.default.a-timer.mean_rate", :value 0.0}])

(deftest map-timer->map
  (let [reg (m/new-registry)
        timer (t/timer reg "a-timer")
        ts (mapcat timer->map (m/timers reg))]
    (is (= (set zero-timer)
           (set ts)))))

(def zero-histo
  [{:type :histogram,
    :name "default.default.a-histogram.p999",
    :value 0.0} 
   {:type :histogram,
    :name "default.default.a-histogram.p50",
    :value 0.0}
   {:type :histogram,
    :name "default.default.a-histogram.p95",
    :value 0.0}
   {:type :histogram,
    :name "default.default.a-histogram.count",
    :value 0}
   {:type :histogram,
    :name "default.default.a-histogram.p75",
    :value 0.0}
   {:type :histogram,
    :name "default.default.a-histogram.min",
    :value 0}
   {:type :histogram,
    :name "default.default.a-histogram.mean",
    :value 0.0}
   {:type :histogram,
    :name "default.default.a-histogram.p99",
    :value 0.0}
   {:type :histogram,
    :name "default.default.a-histogram.max",
    :value 0}
   {:type :histogram,
    :name "default.default.a-histogram.p98",
    :value 0.0}])

(deftest map-histo->map
  (let [reg (m/new-registry)
        histo (h/histogram reg "a-histogram")
        hs (mapcat histo->map (m/histograms reg))]
    (is (= (set zero-histo)
           (set hs)))))

(def zero-meter
  [{:type :meter, :name "default.default.a-meter.m1_rate", :value 0.0}
   {:type :meter,
    :name "default.default.a-meter.mean_rate",
    :value 0.0}
   {:type :meter, :name "default.default.a-meter.m5_rate", :value 0.0}
   {:type :meter,
    :name "default.default.a-meter.m15_rate",
    :value 0.0}])

(deftest map-meter->map
  (let [reg (m/new-registry)
        meter (meter/meter reg "a-meter")
        ms (mapcat meter->map (m/meters reg))]
    (is (= (set zero-meter)
           (set ms)))))

(deftest all-metrics
  (let [reg (m/new-registry)
        meter (meter/meter reg "a-meter")
        histo (h/histogram reg "a-histogram")
        timer (t/timer reg "a-timer")
        gauge (g/gauge-fn reg "a-gauge" (constantly 10))
        counter (c/counter reg "a-counter")
        all-metrcs (registry->maps reg)]
    (is (= (set (map #(assoc % :log-type :metric)
                     (flatten [zero-meter
                               zero-histo
                               zero-timer
                               zero-counter
                               zero-gauge])))
           (set all-metrcs)))))



test/kixi/metrics/name_safety_test.clj
(ns kixi.metrics.name-safety-test
  (:require [kixi.metrics.name-safety :as sut]
            [clojure.test :refer :all]))

(deftest ensure-uri-names-are-clean
  (let [uri "file/08eb375b-bf92-40cc-bced-e7bf74a7ceba/meta"]
    (is (= (sut/safe-name uri)
           "file.GUID.meta"))))


src/kixi/metrics.clj
(ns kixi.metrics
  (:require [cheshire
             [core :as json]
             [factory :as factory]
             [generate :as gen]]
            [metrics
             [core :as core :refer [default-registry]]
             [counters :as c]
             [gauges :as g]
             [histograms :as h]
             [meters :as m]
             [timers :as t]])
  (:import [com.codahale.metrics MetricFilter ScheduledReporter]
           com.fasterxml.jackson.core.JsonFactory
           [java.io BufferedWriter Writer]
           java.util.concurrent.TimeUnit))

(defn gauge->map
  [[gauge-name gauge]]
  {:type :gauge
   :name gauge-name
   :value (g/value gauge)})

(defn counter->map
  [[gauge-name gauge]]
  {:type :counter
   :name gauge-name
   :value (c/value gauge)})

(def sfirst (comp second first))

(def timer-fields
  {"min" t/smallest
   "max" t/largest
   "mean" t/mean
   "std-dev" t/std-dev
   "count" t/number-recorded
   "p50" #(sfirst (t/percentiles % [0.5]))
   "p75" #(sfirst (t/percentiles % [0.75]))
   "p95" #(sfirst (t/percentiles % [0.95]))
   "p98" #(sfirst (t/percentiles % [0.98]))
   "p99" #(sfirst (t/percentiles % [0.99]))
   "p999" #(sfirst (t/percentiles % [0.999]))
   "m1_rate" t/rate-one
   "m5_rate" t/rate-five
   "m15_rate" t/rate-fifteen
   "mean_rate" t/rate-mean})

(defn timer->map
  [[timer-name timer]]
  (map
   (fn [[n f]]
     {:type :timer
      :name (str timer-name "." n)
      :value (f timer)})
   timer-fields))

(def histo-fields
  {"count" h/number-recorded
   "max" h/largest
   "mean" h/mean
   "min" h/smallest
   "p50" #(sfirst (h/percentiles % [0.5]))
   "p75" #(sfirst (h/percentiles % [0.75]))
   "p95" #(sfirst (h/percentiles % [0.95]))
   "p98" #(sfirst (h/percentiles % [0.98]))
   "p99" #(sfirst (h/percentiles % [0.99]))
   "p999" #(sfirst (h/percentiles % [0.999]))})

(defn histo->map
  [[histo-name histo]]
  (map
   (fn [[n f]]
     {:type :histogram
      :name (str histo-name "." n)
      :value (f histo)})
   histo-fields))

(def meter-fields
  {"m1_rate" m/rate-one
   "m5_rate" m/rate-five
   "m15_rate" m/rate-fifteen
   "mean_rate" m/rate-mean})

(defn meter->map
  [[meter-name meter]]
  (map
   (fn [[n f]]
     {:type :meter
      :name (str meter-name "." n)
      :value (f meter)})
   meter-fields))

(def mapper->metric-getter
  {gauge->map core/gauges
   counter->map core/counters
   histo->map core/histograms
   meter->map core/meters
   timer->map core/timers})

(defn registry->maps
  [registry]
  (map 
   #(assoc % :log-type :metric)
   (flatten
    (map
     (fn [[mapper getter]]
       (map mapper (getter registry)))
     mapper->metric-getter))))

(defn get-lock
  [^Writer writer]
  (let [lock-field (.getDeclaredField Writer "lock")]    
    (.setAccessible lock-field true)
    (.get lock-field writer)))

(defn reporter
  ([opts]
   (reporter default-registry opts))
  ([registry {:keys [rate-unit duration-unit filter] :as opts
              :or {filter MetricFilter/ALL
                   rate-unit TimeUnit/SECONDS
                   duration-unit TimeUnit/MILLISECONDS}}]
   (let [name "Json-console-reporter"
         lock (get-lock *out*)]
     (proxy [ScheduledReporter] [registry name filter rate-unit duration-unit]
       (report
         ([]
          (doseq [metric (registry->maps registry)]            
            (locking lock
              (json/generate-stream
               metric
               *out*)
              (prn *out*)))))))))

(defn start
  ([^ScheduledReporter reporter poll]
   (start reporter poll TimeUnit/SECONDS))
  ([^ScheduledReporter reporter poll poll-unit]
   (.start reporter
           poll poll-unit)))


src/kixi/metrics/name_safety.clj
(ns kixi.metrics.name-safety
  (:require [clojure.string :as str]))

;; Lifted all this from silcon-gorge/radix
;; Generally useful for creating metrics for uri's, but good for any dynamic metric name creation.

(def replace-guid
  [#"[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}" "GUID"])

(def replace-number
  [#"[0-9][0-9]+" "NUMBER"])

(def replace-territory
  [#"/[a-z]{2}/" "/TERRITORIES/"])

(defn- apply-regex
  [path [pattern replacement]]
  (str/replace path pattern replacement))

(defn- apply-aggregations
  [path aggregations]
  (reduce apply-regex path aggregations))

;perf: replace with stringbuild and do all these in one pass
(defn clean-metric-name
  [name]
  (-> name
      (str/replace " " ".")
      (str/replace "./" ".")
      (str/replace "/" ".")
      (str/replace #"\p{Cntrl}" "")
      (str/replace #"^\.+" "")
      (str/replace #"\.+$" "")))

(def default-aggregations [replace-guid replace-number replace-territory])

(defn safe-name
  [unsafe-name]
  (when unsafe-name
    (-> unsafe-name
        clean-metric-name
        (apply-aggregations default-aggregations))))


src/kixi/metrics/reporters/json_console.clj
(ns kixi.metrics.reporters.json-console
  (:require [cheshire
             [core :as json]
             [factory :as factory]
             [generate :as gen]]
            [metrics
             [core :as core :refer [default-registry]]
             [counters :as c]
             [gauges :as g]
             [histograms :as h]
             [meters :as m]
             [timers :as t]])
  (:import [com.codahale.metrics MetricFilter ScheduledReporter]
           com.fasterxml.jackson.core.JsonFactory
           [java.io BufferedWriter Writer]
           java.util.concurrent.TimeUnit))

(defn gauge->map
  [[gauge-name gauge]]
  {:type :gauge
   :name gauge-name
   :value (g/value gauge)})

(defn counter->map
  [[gauge-name gauge]]
  {:type :counter
   :name gauge-name
   :value (c/value gauge)})

(def sfirst (comp second first))

(def timer-fields
  {"min" t/smallest
   "max" t/largest
   "mean" t/mean
   "std-dev" t/std-dev
   "count" t/number-recorded
   "p50" #(sfirst (t/percentiles % [0.5]))
   "p75" #(sfirst (t/percentiles % [0.75]))
   "p95" #(sfirst (t/percentiles % [0.95]))
   "p98" #(sfirst (t/percentiles % [0.98]))
   "p99" #(sfirst (t/percentiles % [0.99]))
   "p999" #(sfirst (t/percentiles % [0.999]))
   "m1_rate" t/rate-one
   "m5_rate" t/rate-five
   "m15_rate" t/rate-fifteen
   "mean_rate" t/rate-mean})

(defn timer->map
  [[timer-name timer]]
  (map
   (fn [[n f]]
     {:type :timer
      :name (str timer-name "." n)
      :value (f timer)})
   timer-fields))

(def histo-fields
  {"count" h/number-recorded
   "max" h/largest
   "mean" h/mean
   "min" h/smallest
   "p50" #(sfirst (h/percentiles % [0.5]))
   "p75" #(sfirst (h/percentiles % [0.75]))
   "p95" #(sfirst (h/percentiles % [0.95]))
   "p98" #(sfirst (h/percentiles % [0.98]))
   "p99" #(sfirst (h/percentiles % [0.99]))
   "p999" #(sfirst (h/percentiles % [0.999]))})

(defn histo->map
  [[histo-name histo]]
  (map
   (fn [[n f]]
     {:type :histogram
      :name (str histo-name "." n)
      :value (f histo)})
   histo-fields))

(def meter-fields
  {"m1_rate" m/rate-one
   "m5_rate" m/rate-five
   "m15_rate" m/rate-fifteen
   "mean_rate" m/rate-mean})

(defn meter->map
  [[meter-name meter]]
  (map
   (fn [[n f]]
     {:type :meter
      :name (str meter-name "." n)
      :value (f meter)})
   meter-fields))

(def mapper->metric-getter
  {gauge->map core/gauges
   counter->map core/counters
   histo->map core/histograms
   meter->map core/meters
   timer->map core/timers})

(defn registry->maps
  [registry]
  (map 
   #(assoc % :logtype :metric)
   (flatten
    (map
     (fn [[mapper getter]]
       (map mapper (getter registry)))
     mapper->metric-getter))))

(defn get-lock
  [^Writer writer]
  (let [lock-field (.getDeclaredField Writer "lock")]    
    (.setAccessible lock-field true)
    (.get lock-field writer)))

(defn reporter
  ([opts]
   (reporter default-registry opts))
  ([registry {:keys [rate-unit duration-unit filter] :as opts
              :or {filter MetricFilter/ALL
                   rate-unit TimeUnit/SECONDS
                   duration-unit TimeUnit/MILLISECONDS}}]
   (let [name "Json-console-reporter"
         lock (get-lock *out*)]
     (proxy [ScheduledReporter] [registry name filter rate-unit duration-unit]
       (report
         ([]
          (doseq [metric (registry->maps registry)]            
            (locking lock
              (json/generate-stream
               metric
               *out*)
              (prn)))))))))

(defn start
  ([^ScheduledReporter reporter poll]
   (start reporter poll TimeUnit/SECONDS))
  ([^ScheduledReporter reporter poll poll-unit]
   (.start reporter
           poll poll-unit)))

(defn stop
  [^ScheduledReporter reporter]
  (.stop reporter))
